### PR TITLE
STOR-1065: Publish ClusterRoles for csi driver sidecars

### PR DESCRIPTION
**Pull request by mpatlasov**
The PR adds a bunch of new ClusterRoles to `manifests/` dir. This way those new ClusterRoles will be always created, regardless of cloud. They are building blocks to compose ClusterRoles for CSI dirver sidecars. For example, `external-provisioner` sidecar for `aws-ebs` csi driver can compose the same ClusterRole as in https://github.com/openshift/aws-ebs-csi-driver-operator/blob/master/assets/rbac/provisioner_role.yaml by adding ClusterRoleBindings for `openshift-csi-main-provisioner-role` and `openshift-csi-provisioner-volumesnapshot-reader-role`. The only exception is `leases` rules which need to be moved from ClusterRoles to per-namespace Roles anyway.

As soon as this change is merged into `cluster-storage-operator`, it will be possible to get rid of ClusterRole definitions in csi driver operators, they will only define ClusterRoleBindings referring these new ClusterRoles.

### CODE REVIEW

File: pkg/operator/csidriveroperator/csioperatorclient/aws.go [hunk 1]
```diff
@@ -46,11 +48,15 @@ func GetAWSEBSCSIOperatorConfig(isHypershift bool) CSIOperatorConfig {
...
 			"csidriveroperators/aws-ebs/hypershift/guest/06_clusterrolebinding.yaml",
+			"sidecars/main_provisioner_role.yaml",
+			"sidecars/volumesnapshot_provisioner_role.yaml",
 		}
 		csiDriverConfig.MgmtStaticAssets = []string{
 			"csidriveroperators/aws-ebs/hypershift/mgmt/01_operator_role.yaml",
 			"csidriveroperators/aws-ebs/hypershift/mgmt/02_sa.yaml",
 			"csidriveroperators/aws-ebs/hypershift/mgmt/04_rolebinding.yaml",
+			"sidecars/main_provisioner_role.yaml",
+			"sidecars/volumesnapshot_provisioner_role.yaml",
```
[the remark this answers is not in the document]
Comment (mpatlasov) [reply pkg/operator/csidriveroperator/csioperatorclient/aws.go:59]: We clarified this topic in a chat with Hemant. For posterity: an instance of csi driver running inside Management Cluster communicates with Guest Cluster API server, so ClusterRoles published there (in `csiDriverConfig.GuestStaticAssets` above) would fully suit its needs.

File: assets/sidecars/lease_provisioner_role.yaml [hunk 2]
```diff
@@ -0,0 +1,8 @@
+kind: ClusterRole
+apiVersion: rbac.authorization.k8s.io/v1
+metadata:
+  name: lease-external-provisioner-role
+rules:
+  - apiGroups: ["coordination.k8s.io"]
+    resources: ["leases"]
+    verbs: ["get", "watch", "list", "delete", "update", "create"]
```
Comment (jsafrane) [inline assets/sidecars/lease_provisioner_role.yaml:8]: IMO this is not related to provisioning at all. I'd use some global-sounding role (`csi-driver-lease-leader-election`?) and use just RoleBinding (not ClusterRoleBinding), because the sidecars need these perms. only in namespaces where they run.

File: assets/sidecars/volumesnapshot_provisioner_role.yaml [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+apiVersion: rbac.authorization.k8s.io/v1
+metadata:
+  name: volumesnapshot-external-provisioner-role
+rules:
+  - apiGroups: ["snapshot.storage.k8s.io"]
+    resources: ["volumesnapshots"]
+    verbs: ["get", "list"]
+  - apiGroups: ["snapshot.storage.k8s.io"]
+    resources: ["volumesnapshotcontents"]
+    verbs: ["get", "list"]
```
Comment (jsafrane) [inline assets/sidecars/volumesnapshot_provisioner_role.yaml:11]: dtto, if a storage backed supports snapshots, this (and more) will be part of external-snapshotter role.

File: assets/sidecars/configmap_provisioner_role.yaml [hunk 4]
```diff
@@ -0,0 +1,8 @@
+kind: ClusterRole
+apiVersion: rbac.authorization.k8s.io/v1
+metadata:
+  name: configmap-external-provisioner-role
+rules:
+  - apiGroups: [""]
+    resources: ["configmaps"]
+    verbs: ["get", "list"]
```
[the remark this answers is not in the document]
Comment (mpatlasov) [reply assets/sidecars/configmap_provisioner_role.yaml:8]: https://github.com/openshift/ibm-vpc-block-csi-driver-operator/blob/master/assets/rbac/provisioner_role.yaml#L87-L90

File: assets/sidecars/lease_provisioner_role.yaml (hunk 2 above)
Comment (mpatlasov) [reply assets/sidecars/lease_provisioner_role.yaml:8]: > IMO this is not related to provisioning at all.

We currently have 3 operators defining this role:
https://github.com/openshift/aws-efs-csi-driver-operator/blob/master/assets/rbac/provisioner_role.yaml#L25-L26
https://github.com/openshift/azure-file-csi-driver-operator/blob/main/assets/rbac/provisioner_role.yaml#L25-L26
https://github.com/openshift/ibm-powervs-block-csi-driver-operator/blob/main/assets/rbac/provisioner_role.yaml#L31-L32

File: pkg/operator/csidriveroperator/csioperatorclient/alibaba-disk.go [hunk 5]
```diff
@@ -29,6 +29,8 @@ func GetAlibabaDiskCSIOperatorConfig() CSIOperatorConfig {
 			"csidriveroperators/alibaba-disk/04_rolebinding.yaml",
 			"csidriveroperators/alibaba-disk/05_clusterrole.yaml",
 			"csidriveroperators/alibaba-disk/06_clusterrolebinding.yaml",
+			"sidecars/main_provisioner_role.yaml",
+			"sidecars/volumesnapshot_provisioner_role.yaml",
```
[the remark this answers is not in the document]
Comment (mpatlasov) [reply pkg/operator/csidriveroperator/csioperatorclient/alibaba-disk.go:33]: I like this idea. I'll try to move all roles (for all sidecars AND all operators) into `manifests/` dir under some global-sounding names like `openshift-csi-lease-leader-election-role` and `openshift-csi-secret-reader-role`.

File: assets/sidecars/configmap_provisioner_role.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (mpatlasov) [reply assets/sidecars/configmap_provisioner_role.yaml:8]: @dobsonj You were involved in https://github.com/openshift/ibm-vpc-block-csi-driver-operator/pull/40, and [this comment](https://github.com/openshift/ibm-vpc-block-csi-driver-operator/pull/40/commits/ff4825c94286a4bb9cdcc3d74442fb0e0b3fd6f1#r901363637):

> adding the resources and permissions for provisioner role should suffice. It can be removed from resizer and registrar.

suggests that "configmaps" rule was added intentionally. Do you remember why `external-provisioner` needs it and what makes `ibm-vpc-block-csi-driver-operator` so special that other operators are happy w/o it?

File: assets/sidecars/volumeattachment_provisioner_role.yaml [hunk 6]
```diff
@@ -0,0 +1,8 @@
+kind: ClusterRole
+apiVersion: rbac.authorization.k8s.io/v1
+metadata:
+  name: volumeattachment-external-provisioner-role
+rules:
+  - apiGroups: ["storage.k8s.io"]
+    resources: ["volumeattachments"]
+    verbs: ["get", "list", "watch"]
```
[the remark this answers is not in the document]
Comment (mpatlasov) [reply assets/sidecars/volumeattachment_provisioner_role.yaml:8]: We have to distinguish them because "verbs" lists are different: `["get", "list", "watch"]` for provisioner, and `["get", "list", "watch", "update", "patch"]` for attacher. It would impact security if we let provisioner to modify volumeattachments although it needs only read-only access to them.

File: assets/sidecars/volumesnapshot_provisioner_role.yaml (hunk 3 above)
Comment (mpatlasov) [reply assets/sidecars/volumesnapshot_provisioner_role.yaml:11]: Same as in the comment above: provisioner needs only two verbs: `["get", "list"]`, while snapshotter wants more: `["get", "list", "watch", "update"]`.

File: assets/sidecars/secret_provisioner_role.yaml [hunk 7]
```diff
@@ -0,0 +1,8 @@
+kind: ClusterRole
+apiVersion: rbac.authorization.k8s.io/v1
+metadata:
+  name: secret-external-provisioner-role
+rules:
+  - apiGroups: [""]
+    resources: ["secrets"]
+    verbs: ["get", "list", "watch"]
```
[the remark this answers is not in the document]
Comment (mpatlasov) [reply assets/sidecars/secret_provisioner_role.yaml:8]: verbs lists are slightly different: `["get", "list", "watch"]` for provisioner, and only `["get", "list"]` for snapshotter. I think it's OK to simplify things having a single `openshift-csi-secret-reader.role` with `["get", "list", "watch"]` for all sidecars who wants it.

@dobsonj , from "Threat Model" standpoint, are we supposed to distinguish between `["get", "list"]` and `["get", "list", "watch"]`?

**Comment by mpatlasov** [comment]
The last [force-push](https://github.com/openshift/cluster-storage-operator/pull/379#event-9453548956) fully reworks PR putting all ClusterRole building blocks (for all sidecars) to `manifests/` dir. See this [comment](https://github.com/openshift/cluster-storage-operator/pull/379#discussion_r1219542371).

File: assets/sidecars/secret_provisioner_role.yaml (hunk 7 above)
[the remark this answers is not in the document]
Comment (dobsonj) [reply assets/sidecars/secret_provisioner_role.yaml]: IMO, I think it's acceptable to create a single read-only role (including get/list/watch) for all sidecars to use.

File: assets/sidecars/configmap_provisioner_role.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (dobsonj) [reply assets/sidecars/configmap_provisioner_role.yaml]: AFAIK, [this configmap](https://github.com/openshift/ibm-vpc-block-csi-driver-operator/blob/master/assets/configmap.yaml) is needed by the CSI driver container, but I can't think of a reason why external-provisioner would need it.

File: assets/sidecars/volumeattachment_provisioner_role.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (jsafrane) [reply assets/sidecars/volumeattachment_provisioner_role.yaml]: The provisioner and attacher use the same service account, so there is no impact on security. IMO having the permissions that detailed will impact readability.

[the remark this answers is not in the document]
Comment (mpatlasov) [reply assets/sidecars/volumeattachment_provisioner_role.yaml]: @jsafrane , how can we apply your explanation about provisioner and attacher to the manila usecase? We don't have `attacher_role.yaml` there, but in `provisioner_role.yaml`:

```
  - apiGroups: ["storage.k8s.io"]
    resources: ["volumeattachments"]
    verbs: ["get", "list", "watch"]
```
(https://github.com/openshift/csi-driver-manila-operator/blob/master/assets/rbac/provisioner_role.yaml#L24-L26)

**Comment by mpatlasov** [comment]
New force-push updates PR:
1) remove `manifests/openshift-csi-csinodeinfo-reader-role.yaml` altogether (per discussion with Jan)
2) added annotations to make CVO happy

File: manifests/sidecar-main_attacher_role.yaml [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+    include.release.openshift.io/self-managed-high-availability: "true"
+    include.release.openshift.io/single-node-developer: "true"
+    capability.openshift.io/name: Storage
+rules:
+  - apiGroups: [""]
+    resources: ["persistentvolumes"]
+    verbs: ["get", "list", "watch", "update", "patch"]
+  - apiGroups: [""]
+    resources: ["nodes"]
+    verbs: ["get", "list", "watch"]
```
Comment (jsafrane) [inline manifests/sidecar-main_attacher_role.yaml:16]: I would expect some rules for `volumeattachment` - the attacher reads/writes them a lot

File: assets/sidecars/volumeattachment_provisioner_role.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (jsafrane) [reply assets/sidecars/volumeattachment_provisioner_role.yaml]: It looks like a bug in Manila then.

[the remark this answers is not in the document]
Comment (jsafrane) [reply assets/sidecars/volumeattachment_provisioner_role.yaml]: or, maybe not...  Keep the reader role there, we'll see when we get to individual drivers.

File: manifests/sidecar-main_attacher_role.yaml (hunk 8 above)
Comment (mpatlasov) [reply manifests/sidecar-main_attacher_role.yaml:16]: Yes, that's right. The attacher will bind to openshift-csi-volumeattachment-reader-role and openshift-csi-volumeattachment-writer-role. I didn't include these rules into `openshift-csi-main-attacher-role` to avoid verb duplication.

**Comment by mpatlasov** [comment]
New force-push updates PR:
 * add a rule for `csinodes` to `sidecar-main_attacher_role.yaml`
 * remove `openshift-csi-csinode-reader-role.yaml` altogether

This simplifies the code because all `attacher` sidecars need prems to read csinodes.

**Comment by mpatlasov** [comment]
New force-push updates PR:
 * added rules to `manifests/sidecar-main_attacher_role.yaml`. Now it suites all operators
 * reworked `manifests/openshift-csi-*` files to correspond to specific sidecars
 
 (see https://docs.google.com/document/d/1oiGN7ulFM3BE-fPb6U2mbSWbAQCJYyFJm3CYzHrkKkY/edit?usp=sharing)

**Comment by mpatlasov** [comment]
This PR passes e2e tests in conjunction with [STOR-1065: Publish ClusterRoles for csi driver sidecars](https://github.com/openshift/cluster-storage-operator/pull/379):
```
job <https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/release-openshift-origin-installer-launch-aws-modern/1671702027480076288 | test e2e https://github.com/openshift/cluster-storage-operator/pull/379,https://github.com/openshift/aws-ebs-csi-driver-operator/pull/244 aws> succeeded
```
(https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/release-openshift-origin-installer-launch-aws-modern/1671702027480076288)

File: manifests/openshift-csi-resizer-storageclass-reader-role.yaml [hunk 9]
```diff
@@ -0,0 +1,13 @@
+kind: ClusterRole
+apiVersion: rbac.authorization.k8s.io/v1
+metadata:
+  name: openshift-csi-resizer-storageclass-reader-role
```
Comment (jsafrane) [inline manifests/openshift-csi-resizer-storageclass-reader-role.yaml:4]: This role is suspicious - is there a resizer that does not need to read StorageClasses?

File: manifests/sidecar-main_registrar_role.yaml [hunk 10]
```diff
@@ -0,0 +1,32 @@
+kind: ClusterRole
+apiVersion: rbac.authorization.k8s.io/v1
+metadata:
+  name: openshift-csi-main-driver-registrar-role
```
Comment (jsafrane) [inline manifests/sidecar-main_registrar_role.yaml:4]: In https://github.com/openshift/ibm-vpc-block-csi-driver-operator/pull/63 I checked that these rules are ibm-vpc-node-label-updater rules, not registrar's. Node-driver-registrar does not talk to the API server at all.

This ClusterRole (or Role?) belongs to IBM VPC CSI driver operator only.

**Review by jsafrane — COMMENTED** [review]
I am fine if it's merged as it is, we can always fix it later. Still, there is one issue below.

File: manifests/sidecar-main_registrar_role.yaml (hunk 10 above)
Comment (mpatlasov) [reply manifests/sidecar-main_registrar_role.yaml]: OK, `manifests/sidecar-main_registrar_role.yaml` is removed!
